### PR TITLE
Blazor RCL support and shared MAUI services

### PR DESCRIPTION
**Pull request by Eilon**
Several major changes in this PR:
- BlazorWebView now uses the existing IServiceProvider that is available in MAUI
	- You must enable MS.Extensions.DI to use it
	- @PureWeen - could you review this part of the change (fairly small)

- BlazorWebView initialization now typically occurs before the first RootComponent is added, so it uses change notification to add/remove RootComponents as the list changes
   - @SteveSandersonMS - this should look familiar to you because the code is now mostly copied from the WPF/WinForms versions for listening to root component changes
- Razor Class Library (RCL) static assets support
	- The NUPKG contains build props/targets that use the Static Web Assets feature of Razor to gather and package the static assets
	- The targets take all the Razor Static Web Assets and embed them (yuck) in the project assembly as EmbeddedResources, along with an EmbeddedManfiestFileProvider manifest (to preserve folder structure and filenames)
	- Platforms:
		- Windows (WinUI) is **fully implemented**
		- Android is **fully implemented**
		- iOS/MacCatalyst is **fully implemented**
	- @javiercn - this part is for you

- The new targets also include some workarounds for various issues:
	- Work around Razor components ("Content") being detected by the iOS/MacCatalyst SDK as BundleResources (workaround: Mark Razor components as RazorComponent)
	- Work around static assets ("Content") being detected by the iOS/MacCatalyst SDK as BundleResources (workaround: Temporarily remove wwwroot assets from Content, then add them back later)
	- @javiercn - this part is also for you

### CODE REVIEW

File: src/BlazorWebView/src/core/Android/AndroidAssetFileProvider.cs
Comment (SteveSandersonMS) [inline src/BlazorWebView/src/core/Android/AndroidAssetFileProvider.cs:81]: Does this read all the existing static assets into memory on every app startup? If so it seems like it might be a bit inefficient and contribute to delaying the app startup, particularly for any apps that have any large static content items.

Not saying it should be blocking for a preview, but it might be worth tracking some plan to put an overall checksum in place - across all the static assets - and then just comparing that single top-level checksum so we know whether a new version has been installed since the last startup.

File: src/BlazorWebView/src/core/Android/BlazorWebViewHandler.Android.cs
Comment (SteveSandersonMS) [inline src/BlazorWebView/src/core/Android/BlazorWebViewHandler.Android.cs:86]: There might be a race condition here. If L86 runs before the dispatcher continuation from L80, and the collection changes in between those two executions, then I think a root component would be added twice.

In practice it probably wouldn't happen because we're probably already on the dispatcher thread, but just in case, maybe it's worth *not* dispatching, and instead using `Dispatcher.CheckAccess()` (throwing if false) to be sure we didn't need to.

File: src/BlazorWebView/src/core/Windows/BlazorWebViewHandler.Windows.cs
Comment (SteveSandersonMS) [inline src/BlazorWebView/src/core/Windows/BlazorWebViewHandler.Windows.cs:62]: Same potential race condition here

File: src/BlazorWebView/src/core/iOS/BlazorWebViewHandler.iOS.cs
Comment (SteveSandersonMS) [inline src/BlazorWebView/src/core/iOS/BlazorWebViewHandler.iOS.cs:133]: Also race here (or again, maybe not, depending on whether we're already on the dispatcher)

File: src/BlazorWebView/src/core/iOS/BlazorWebViewHandler.iOS.cs
Comment (SteveSandersonMS) [inline src/BlazorWebView/src/core/iOS/BlazorWebViewHandler.iOS.cs:201]: The amount of duplication across the three hosting flavours here makes me wonder if something common could be extracted from all three. Not suggesting you should hold up this PR for that, but might save time in the long run.

File: src/Controls/samples/Controls.Sample.SingleProject/Startup.cs
Comment (SteveSandersonMS) [inline src/Controls/samples/Controls.Sample.SingleProject/Startup.cs:23]: There are some pretty eclectic variations in indentation across different files in this PR. Is it because the MAUI repo as a whole has differing conventions in different places?

But this `{` in particular is a true maverick.

File: src/Controls/samples/Controls.Sample.WinUI/Maui.Controls.Sample.WinUI.csproj
Comment (SteveSandersonMS) [inline src/Controls/samples/Controls.Sample.WinUI/Maui.Controls.Sample.WinUI.csproj:29]: Rather than adding these in different places manually, would it work to add `Directory.Build.props` and `Directory.Build.targets` files?

Also is there any symbol that could be used in place of the `net6.0-windows`? We know for sure that version number is going to get stale.

File: src/Controls/samples/Controls.Sample/Maui.Controls.Sample-net6.csproj
Comment (SteveSandersonMS) [inline src/Controls/samples/Controls.Sample/Maui.Controls.Sample-net6.csproj:59]: Should these be here? Doesn't the SDK do this automatically for files in `wwwroot`?

File: src/Controls/samples/Controls.Sample/Startup.cs
Comment (SteveSandersonMS) [inline src/Controls/samples/Controls.Sample/Startup.cs:3]: I must be missing something here. When would `NET6_0_OR_GREATER` ever not be defined? That's all we support here, isn't it?

File: src/Controls/samples/Controls.Sample/Startup.cs
Comment (SteveSandersonMS) [inline src/Controls/samples/Controls.Sample/Startup.cs:99]: It will be interested to see the project template. I deeply hope it doesn't look anything like this file, which is mystifying.

File: src/Controls/samples/Controls.Sample/wwwroot/index.html
Comment (javiercn) [inline src/Controls/samples/Controls.Sample/wwwroot/index.html:9]: This is wrong I believe. You should reference `Controls.Sample.styles.css`, the `.bundle.scp.css` files only contains the bundle for the scoped css files, it doesn't have import references to other libraries. This works right now because you don't have transitive RCLs in the app.

File: src/Controls/samples/Controls.Sample.Droid/Maui.Controls.Sample.Droid-net6.csproj
Comment (javiercn) [inline src/Controls/samples/Controls.Sample.Droid/Maui.Controls.Sample.Droid-net6.csproj:9]: Hmm, not sure I understand why this is needed, it should be enabled based on the fact you are referencing the SDK

File: src/Controls/samples/Controls.Sample/Startup.cs
Comment (rmarinho) [reply src/Controls/samples/Controls.Sample/Startup.cs:3]: we build this sln in a not .net6 world also for now (legacy to better have tool support)

File: src/Controls/samples/Controls.Sample.Droid/Maui.Controls.Sample.Droid-net6.csproj
Comment (javiercn) [inline src/Controls/samples/Controls.Sample.Droid/Maui.Controls.Sample.Droid-net6.csproj:44]: Is this something that will appear on the template (almost sure the answer is no, but have to ask 😄 )

File: src/BlazorWebView/src/core/iOS/BlazorWebViewHandler.iOS.cs
Comment (javiercn) [inline src/BlazorWebView/src/core/iOS/BlazorWebViewHandler.iOS.cs:191]: Can this be a problem here? I'm not sure if we've defined equality for the root component type.

File: src/BlazorWebView/src/core/Windows/BlazorWebViewHandler.Windows.cs
Comment (javiercn) [inline src/BlazorWebView/src/core/Windows/BlazorWebViewHandler.Windows.cs:83]: What's the reason for using `global::` here?

**Review by javiercn — COMMENTED** [review]
I've made a few comments, things look good in general, just small fixes.

File: src/BlazorWebView/src/core/Android/AndroidAssetFileProvider.cs
Comment (mattleibow) [reply src/BlazorWebView/src/core/Android/AndroidAssetFileProvider.cs:81]: Or maybe it is possible to just read directly from the assets files? There are some APIs and we could potentially just read the zip entries further.

Comment (Eilon) [reply src/BlazorWebView/src/core/Android/AndroidAssetFileProvider.cs:81]: Discussed with @Redth and @javiercn this morning and we will use a different approach for static assets using `MauiAssets`. This will all go away.

File: src/Controls/samples/Controls.Sample.SingleProject/Startup.cs
Comment (Eilon) [reply src/Controls/samples/Controls.Sample.SingleProject/Startup.cs:23]: The MAUI bot will fix my mess later 😄

File: src/Controls/samples/Controls.Sample/Startup.cs
Comment (Eilon) [reply src/Controls/samples/Controls.Sample/Startup.cs:3]: Right, this shared sample is currently used also in non-.NET 6 scenarios, so all Blazor stuff is excluded from such builds. I think this eventually goes away and it's all .NET 6, but for now it needs to stay here to compile.

File: src/Controls/samples/Controls.Sample/Startup.cs
Comment (Eilon) [reply src/Controls/samples/Controls.Sample/Startup.cs:99]: The project templates will be .NET 6 only so all the `#if`s go away.

File: src/Controls/samples/Controls.Sample/wwwroot/index.html
Comment (Eilon) [reply src/Controls/samples/Controls.Sample/wwwroot/index.html:9]: For others watching: I met with @javiercn and @Redth today and we have a different solution that should fix this.

File: src/BlazorWebView/src/core/Windows/BlazorWebViewHandler.Windows.cs
Comment (Eilon) [reply src/BlazorWebView/src/core/Windows/BlazorWebViewHandler.Windows.cs]: To avoid a conflict with some sub-namespace that has `System` in it. But I think this code is gone now.

File: src/Controls/samples/Controls.Sample.Droid/Maui.Controls.Sample.Droid-net6.csproj
Comment (Eilon) [reply src/Controls/samples/Controls.Sample.Droid/Maui.Controls.Sample.Droid-net6.csproj:44]: It won't 😄

File: src/BlazorWebView/src/core/MauiAssetFileProvider.Windows.cs
Comment (Eilon) [inline src/BlazorWebView/src/core/MauiAssetFileProvider.Windows.cs:21]: @Redth - FYI the current Resizetizer stuff seems to remove folder paths, so if I have a file `wwwroot\css\main\foo.css` the file ends up in `Appx\Assets\foo.css` and I lose the sub-folders. We need a way to preserve the folder structure.

(I tried hacking up some MSBuild goop to do it but couldn't figure it out.)

Do you have any concerns with adding support for sub-folders in MauiAssets?

File: src/BlazorWebView/src/core/MauiAssetFileProvider.Windows.cs
Comment (Eilon) [inline src/BlazorWebView/src/core/MauiAssetFileProvider.Windows.cs:23]: @Redth - also FYI on this. Using the package APIs, which are all async, e.g. `GetItemAsync()`, would all throw mysterious C++ exceptions. I wonder if I'm running it on the wrong thread? (Which would be annoying, because nothing else here is async, and calling `GetResult()` is awful.)

Did you ever see any exceptions while using those APIs?

File: src/BlazorWebView/src/core/build/Microsoft.AspNetCore.Components.WebView.Maui.targets
Comment (Eilon) [inline src/BlazorWebView/src/core/build/Microsoft.AspNetCore.Components.WebView.Maui.targets:4]: @javiercn - do your StaticWebAsset magic here. I believe it's the right place.

**Comment by Eilon** [comment]
For anyone looking, I'm focusing only on WinUI working right now. Once that's ship-shape it'll make bringing up other platforms much easier because almost all the static asset code will be platform-independent shared code. Only the MauiAssetFileProvider will be platform-dependent.

File: src/BlazorWebView/src/core/iOS/BlazorWebViewHandler.iOS.cs
Comment (SteveSandersonMS) [reply src/BlazorWebView/src/core/iOS/BlazorWebViewHandler.iOS.cs:191]: I would have thought the default struct equality rules would have the right behavior in this case.

File: src/BlazorWebView/src/core/Android/AndroidAssetFileProvider.cs
Comment (Eilon) [reply src/BlazorWebView/src/core/Android/AndroidAssetFileProvider.cs:81]: I'm putting the Android work on hold for now because it's not a high priority for this preview. I'll leave this code here anyway.

File: src/BlazorWebView/src/core/iOS/BlazorWebViewHandler.iOS.cs
Comment (Eilon) [reply src/BlazorWebView/src/core/iOS/BlazorWebViewHandler.iOS.cs]: It's the same exact code from WPF/WinForms. I'll log an issue to confirm.

File: src/BlazorWebView/src/core/iOS/BlazorWebViewHandler.iOS.cs
Comment (Eilon) [reply src/BlazorWebView/src/core/iOS/BlazorWebViewHandler.iOS.cs]: I've cleaned it up and put the common code in a common file.

File: src/BlazorWebView/src/core/Android/BlazorWebViewHandler.Android.cs
Comment (Eilon) [reply src/BlazorWebView/src/core/Android/BlazorWebViewHandler.Android.cs]: I think I've fixed this now (but the code is in another file now). But if equality works the way we think, I think the logic in the CollectionChanged handler would avoid any potential problem anyway, no?

**Comment by Eilon** [comment]
STATUS: I have confirmed that WinUI and Android are both functional. I'd like to get these checked in ASAP to unblock other work.

@javiercn / @SteveSandersonMS - please let me know if you have any concerns on the Blazor-isms.
@PureWeen - there's not much MAUI-isms here, but I'd appreciate a look all the same.

**Review by javiercn — APPROVED** [review]

File: src/Controls/samples/Controls.Sample.Droid/Maui.Controls.Sample.Droid-net6.csproj
Comment (Redth) [inline src/Controls/samples/Controls.Sample.Droid/Maui.Controls.Sample.Droid-net6.csproj:1]: @jonathanpeppers how do we make this not needed, and instead import from a property?

Comment (jonathanpeppers) [reply src/Controls/samples/Controls.Sample.Droid/Maui.Controls.Sample.Droid-net6.csproj:1]: So if we followed the pattern of what `$(UseMaui)` will eventually be, maybe we should come up with a new property like `$(UseRazor)`?

I think we could do something like:

```xml
<PropertyGroup>
  <_RazorSdkImportsMicrosoftNetSdk  Condition=" '$(UseRazor)' == 'true' ">false</_RazorSdkImportsMicrosoftNetSdk>
</PropertyGroup>
<Import Condition=" '$(UseRazor)' == 'true' " Sdk="Microsoft.NET.Sdk.Razor" Project="Sdk.props" />
```
And put this in `Microsoft.Maui.Controls.props` ^^

We'd also need to import in `Microsoft.Maui.Controls.targets`:

```xml
<Import Condition=" '$(UseRazor)' == 'true' " Sdk="Microsoft.NET.Sdk.Razor" Project="Sdk.targets" />
```

Then any project that has `Microsoft.NET.Sdk.Razor` would define `<UseRazor>true</UseRazor>` instead.

File: src/Core/src/Hosting/AppHostBuilderExtensions.cs
Comment (Eilon) [inline src/Core/src/Hosting/AppHostBuilderExtensions.cs:105]: @PureWeen / @Redth / @rmarinho - any problem with adding this extension method? Hopefully we figure out a way to avoid this entirely, but for preview4, how's this?

**Comment by Eilon** [comment]
OK folks, I think I finally have this stuff figured out. There's a bit of duct tape in there, but it's entirely hidden from the user's projects, meaning we can fix much of it post preview4 (and discussions are underway for many of those issues).

I tested in these scenarios:
- Project type: SingleProject vs. Multi-project
- Platform: WinUI, Android, iOS, MacCatalyst
- Asset location: All static assets/Razor files in main project vs. some assets/Razor files coming from Razor Class Libraries (RCL)

I didn't test every combinatorial scenario, but I did cover most.

I would appreciate a review ASAP (the code is much less than before!) so I can unblock the remaining Blazor work (update dependencies in #933, and then project templates).

**Review by d066hie — APPROVED** [review]

**Comment by Eilon** [comment]
Except one thing, I 100% messed up a rebase, so fixing that 😢

**Comment by Eilon** [comment]
For numerous reasons, I'm closing this PR and I opened a new PR here: https://github.com/dotnet/maui/pull/1021

**Pull request closed without merging**
